File: client-pc/daemon-linux/internal/reporter/reporter.go
package reporter

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"sort"
	"time"

	"packetpilot-daemon/internal/logger"
	"packetpilot-daemon/internal/monitor"
)

type Reporter struct {
	config     *Config
	logger     *logger.Logger
	httpClient *http.Client
	monitor    *monitor.TrafficMonitor
}

type Config struct {
	ServerHost     string
	ServerPort     int
	UseTLS         bool
	APIKey         string
	DeviceID       string
	ReportInterval time.Duration
	BatchSize      int
	RetryAttempts  int
	RetryDelay     time.Duration
}

type InterfaceUsageReport struct {
	Interface string  `json:"interface"`
	TotalRx   uint64  `json:"total_rx"`   // bytes
	TotalTx   uint64  `json:"total_tx"`   // bytes
	TotalRxMB float64 `json:"total_rx_mb"` // MB
	TotalTxMB float64 `json:"total_tx_mb"` // MB
}

type DailyUsageReport struct {
	DeviceID    string                   `json:"device_id"`
	Timestamp   time.Time                `json:"timestamp"`
	Date        string                   `json:"date"`        // YYYY-MM-DD
	Interfaces  []InterfaceUsageReport   `json:"interfaces"`
	TotalRxMB   float64                  `json:"total_rx_mb"` // Combined MB
	TotalTxMB   float64                  `json:"total_tx_mb"` // Combined MB
}

type ServerResponse struct {
	Success  bool      `json:"success"`
	Message  string    `json:"message"`
	Commands []Command `json:"commands,omitempty"`
}

type Command struct {
	Type    string `json:"type"`
	AppName string `json:"app_name"`
	Action  string `json:"action"`
}

func NewReporter(config *Config, logger *logger.Logger, monitor *monitor.TrafficMonitor) *Reporter {
	return &Reporter{
		config: config,
		logger: logger,
		httpClient: &http.Client{Timeout: 30 * time.Second},
		monitor: monitor,
	}
}

func (r *Reporter) Start(ctx context.Context) error {
	r.logger.Info("Starting daily usage reporter",
		"server", fmt.Sprintf("%s:%d", r.config.ServerHost, r.config.ServerPort),
		"interval", r.config.ReportInterval,
	)

	ticker := time.NewTicker(r.config.ReportInterval)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			if err := r.sendReport(); err != nil {
				r.logger.Error("Failed to send report", "error", err)
			}
		case <-ctx.Done():
			r.logger.Info("Daily usage reporter stopped")
			return nil
		}
	}
}

func (r *Reporter) sendReport() error {
	// Get current daily usage from monitor
	usage := r.monitor.GetDailyUsage()
	if usage == nil {
		r.logger.Debug("No daily usage data to report")
		return nil
	}

	r.logger.Debug("Got daily usage data", "interfaces_count", len(usage.Interfaces), "date", usage.Date)

	// Create interface reports (sorted by interface name)
	var interfaceReports []InterfaceUsageReport
	var totalRx, totalTx uint64

	// Get sorted interface names
	interfaceNames := make([]string, 0, len(usage.Interfaces))
	for name := range usage.Interfaces {
		interfaceNames = append(interfaceNames, name)
	}
	sort.Strings(interfaceNames)

	// Create reports in sorted order
	for _, iface := range interfaceNames {
		stats := usage.Interfaces[iface]
		interfaceReport := InterfaceUsageReport{
			Interface: iface,
			TotalRx:   stats.TotalRx,
			TotalTx:   stats.TotalTx,
			TotalRxMB: float64(stats.TotalRx) / (1024 * 1024),
			TotalTxMB: float64(stats.TotalTx) / (1024 * 1024),
		}
		interfaceReports = append(interfaceReports, interfaceReport)
		
		// Add to totals
		totalRx += stats.TotalRx
		totalTx += stats.TotalTx
		
		// r.logger.Debug("Added interface to report", "interface", iface, "rx_mb", interfaceReport.TotalRxMB, "tx_mb", interfaceReport.TotalTxMB)
	}

	r.logger.Debug("Created interface reports", "count", len(interfaceReports), "total_rx_mb", float64(totalRx)/(1024*1024), "total_tx_mb", float64(totalTx)/(1024*1024))

	// Create report
	report := DailyUsageReport{
		DeviceID:   r.config.DeviceID,
		Timestamp:  time.Now(),
		Date:       usage.Date,
		Interfaces: interfaceReports,
		TotalRxMB:  float64(totalRx) / (1024 * 1024),
		TotalTxMB:  float64(totalTx) / (1024 * 1024),
	}

	jsonData, err := json.Marshal(report)
	if err != nil {
		return fmt.Errorf("failed to marshal report: %w", err)
	}

	protocol := "http"
	if r.config.UseTLS {
		protocol = "https"
	}
	url := fmt.Sprintf("%s://%s:%d/api/v1/traffic/report", protocol, r.config.ServerHost, r.config.ServerPort)

	var lastErr error
	for attempt := 1; attempt <= r.config.RetryAttempts; attempt++ {
		// Recreate request each attempt so Body isn't exhausted
		req, err := http.NewRequest("POST", url, bytes.NewReader(jsonData))
		if err != nil {
			return fmt.Errorf("failed to create request: %w", err)
		}
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("Authorization", "Bearer "+r.config.APIKey)
		req.Header.Set("User-Agent", "PacketPilot-Daemon/1.0")

		r.logger.Debug("Sending daily usage report", "attempt", attempt, "url", url,
			"date", report.Date, "interfaces", len(report.Interfaces), "total_rx_mb", report.TotalRxMB, "total_tx_mb", report.TotalTxMB)

		resp, err := r.httpClient.Do(req)
		if err != nil {
			lastErr = fmt.Errorf("request failed: %w", err)
			r.logger.Warn("Report attempt failed", "attempt", attempt, "error", err)
			if attempt < r.config.RetryAttempts {
				time.Sleep(r.config.RetryDelay)
				continue
			}
			break
		}

		func() {
			defer resp.Body.Close()
			if resp.StatusCode != http.StatusOK {
				lastErr = fmt.Errorf("server returned status %d", resp.StatusCode)
				r.logger.Warn("Server returned error status", "status", resp.StatusCode)
				return
			}

			var serverResp ServerResponse
			if err := json.NewDecoder(resp.Body).Decode(&serverResp); err != nil {
				r.logger.Warn("Failed to parse server response", "error", err)
			} else {
				r.logger.Info("Daily usage report sent successfully", 
					"message", serverResp.Message,
					"date", report.Date,
					"interfaces", len(report.Interfaces),
					"total_rx_mb", report.TotalRxMB,
					"total_tx_mb", report.TotalTxMB)
				if len(serverResp.Commands) > 0 {
					r.handleCommands(serverResp.Commands)
				}
			}
		}()

		// Success - return (don't reset stats, keep accumulating daily usage)
		return nil
	}

	return lastErr
}

func (r *Reporter) handleCommands(commands []Command) {
	r.logger.Info("Received commands from server", "count", len(commands))
	for _, cmd := range commands {
		r.logger.Info("Processing command", "type", cmd.Type, "app", cmd.AppName, "action", cmd.Action)
		switch cmd.Type {
		case "block_app":
			r.handleBlockApp(cmd.AppName, cmd.Action)
		case "limit_app":
			r.handleLimitApp(cmd.AppName, cmd.Action)
		default:
			r.logger.Warn("Unknown command type", "type", cmd.Type)
		}
	}
}

func (r *Reporter) handleBlockApp(appName, action string) {
	r.logger.Info("Block app command", "app", appName, "action", action)
}

func (r *Reporter) handleLimitApp(appName, action string) {
	r.logger.Info("Limit app command", "app", appName, "action", action)
}

File: client-pc/daemon-win/PacketPilot.Daemon.Win/Reporter/Reporter.cs
using PacketPilot.Daemon.Win.Config;
using PacketPilot.Daemon.Win.Logger;
using PacketPilot.Daemon.Win.Monitor;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using PacketPilot.Daemon.Win.Utils;
using System.Threading;
using System.Threading.Tasks;
using PacketPilot.Daemon.Win.Models;
namespace PacketPilot.Daemon.Win.Reporter
{
    public class Reporter
    {
        private readonly ReporterConfig _config;
        private readonly Logger.Logger _logger;
        private readonly HttpClient _httpClient;
        private readonly ProcessNetworkMonitor _processMonitor;
        private readonly TrafficMonitor _trafficMonitor;
        private readonly ReporterMode _reportMode;
        private CancellationTokenSource? _cancellationTokenSource;

        public Reporter(ReporterConfig config, Logger.Logger logger, ProcessNetworkMonitor processMonitor, TrafficMonitor trafficMonitor, ReporterMode reportMode = ReporterMode.TotalUsage)
        {
            _config = config;
            _logger = logger;
            // TODO: Create strategy class for different modes
            _processMonitor = processMonitor;
            _trafficMonitor = trafficMonitor;
            _reportMode = reportMode;
            _httpClient = new HttpClient { Timeout = UtilsHelper.ParseTimeSpan(_config.ReportInterval) };
        }

        public async Task StartAsync(CancellationToken cancellationToken)
        {
            _cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);

            _logger.Info("Starting daily usage reporter",
                "server", $"{_config.ServerHost}:{_config.ServerPort}",
                "interval", _config.ReportInterval);

            // Send initial health check when starting
            await SendHealthCheckRequest();

            var interval = UtilsHelper.ParseTimeSpan(_config.ReportInterval);
            try
            {
                await AsyncTask.RunPeriodicAsync(
                    interval,
                    async token =>
                    {
                        try
                        {
                            await SendReportAsync().ConfigureAwait(false);
                        }
                        catch (Exception ex)
                        {
                            _logger.Error("Failed to send report", "error", ex.ToString());
                        }
                    },
                    _cancellationTokenSource.Token);
            }
            catch (OperationCanceledException)
            {
                _logger.Info("Daily usage reporter stopped");
            }
        }

        public void Stop()
        {
            _cancellationTokenSource?.Cancel();
            _cancellationTokenSource?.Dispose();
        }

        private async Task<bool> SendHealthCheckRequest()
        {
            var protocol = _config.UseTls ? "https" : "http";
            var url = $"{protocol}://{_config.ServerHost}:{_config.ServerPort}/api/v1/device/health-check";

            try
            {
                using var request = new HttpRequestMessage(HttpMethod.Post, url);
                request.Headers.Add("Authorization", $"Bearer {_config.ApiKey}");
                request.Headers.Add("User-Agent", "PacketPilot-Windows-Daemon/1.0");

                _logger.Debug("Sending health check", "url", url);

                using var response = await _httpClient.SendAsync(request, _cancellationTokenSource?.Token ?? CancellationToken.None);

                if (response.IsSuccessStatusCode)
                {
                    var responseContent = await response.Content.ReadAsStringAsync();
                    _logger.Info("Health check sent successfully", "message", responseContent);
                    return true;
                }
                else
                {
                    _logger.Warn("Health check failed", "status", (int)response.StatusCode);
                    return false;
                }
            }
            catch (Exception ex)
            {
                _logger.Error("Health check error", "error", ex.Message);
                return false;
            }
        }

        private async Task SendReportAsync()
        {
            if (_reportMode == ReporterMode.TotalUsage)
            {
                await CreateTotalUsagePayload();
            }
            else
            {
                await CreatePerProcessUsagePayload();
            }
        }

        private async Task CreatePerProcessUsagePayload()
        {
            var processUsage = _processMonitor.GetProcessUsage();

            if (processUsage == null || processUsage.Count == 0)
            {
                _logger.Debug("No usage data to report");
                return;
            }
            var usageApps = new Dictionary<string, AppUsage>();

            foreach (var kvp in processUsage)
            {
                var identifier = kvp.Key;
                var process = kvp.Value;

                usageApps[identifier] = new AppUsage
                {
                    Identifier = identifier,
                    DisplayName = process.ProcessName,
                    IconHash = process.ProcessIconBase64,
                    TotalRx = process.TotalRxBytes,
                    TotalTx = process.TotalTxBytes,
                    LastSeen = process.LastSeen
                };
            }

            if (usageApps == null || usageApps.Count == 0)
            {
                _logger.Debug("No usage data to report");
                return;
            }

            // Register or update apps first
            var appsToRegister = usageApps.Values
                .Select(app => new AppRegistration
                {
                    Identifier = app.Identifier,
                    DisplayName = string.IsNullOrEmpty(app.DisplayName) ? null : app.DisplayName,
                    IconHash = string.IsNullOrEmpty(app.IconHash) ? null : app.IconHash
                })
                .ToList();

            if (appsToRegister.Count > 0)
            {
                await RegisterAppsRequest(appsToRegister);
            }

            // Send usage reports
            long totalRx = 0, totalTx = 0;
            var appUsageData = new List<AppUsageData>();

            foreach (var kvp in usageApps.OrderBy(x => x.Key))
            {
                var app = kvp.Value;

                appUsageData.Add(new AppUsageData
                {
                    Identifier = app.Identifier,
                    TotalRx = app.TotalRx,
                    TotalTx = app.TotalTx
                });

                // Add to totals
                totalRx += app.TotalRx;
                totalTx += app.TotalTx;
            }

            _logger.Debug("Created app usage data", "count", appUsageData.Count, "total_rx_mb", (double)totalRx / (1024 * 1024), "total_tx_mb", (double)totalTx / (1024 * 1024));

            // Create usage report (only identifiers and usage data)
            var today = DateTime.UtcNow.ToString("yyyy-MM-dd");
            var usageReport = new UsageReportRequest
            {
                Timestamp = DateTime.UtcNow.ToString("o"),
                Date = today,
                Apps = appUsageData
            };

            await SendPerProcessReportRequest(usageReport, today, appUsageData.Count, totalRx, totalTx);
        }

        private async Task CreateTotalUsagePayload()
        {
            var hourlyUsage = _trafficMonitor.GetTotalUsage();
            if (hourlyUsage == null || hourlyUsage.Interfaces.Count == 0)
            {
                _logger.Debug("No total usage data to report");
                return;
            }

            long totalRx = 0;
            long totalTx = 0;

            foreach (var iface in hourlyUsage.Interfaces.Values)
            {
                totalRx += iface.TotalRx;
                totalTx += iface.TotalTx;
            }

            if (totalRx == 0 && totalTx == 0)
            {
                _logger.Debug("Total usage counters are zero, skipping report");
                return;
            }

            var today = DateTime.UtcNow.ToString("yyyy-MM-dd");
            var totalUsageReport = new TotalUsageReportRequest
            {
                Timestamp = DateTime.UtcNow.ToString("o"),
                Date = today,
                TotalRx = totalRx,
                TotalTx = totalTx
            };

            await SendTotalUsageReportRequest(totalUsageReport, today, totalRx, totalTx);
        }

        private async Task RegisterAppsRequest(List<AppRegistration> apps)
        {
            var protocol = _config.UseTls ? "https" : "http";
            var url = $"{protocol}://{_config.ServerHost}:{_config.ServerPort}/api/v1/traffic/register-apps";

            var appsPayload = new RegisterAppsRequest
            {
                Apps = apps
            };

            var jsonData = JsonSerializer.Serialize(appsPayload);

            try
            {
                using var content = new StringContent(jsonData, Encoding.UTF8, "application/json");
                using var request = new HttpRequestMessage(HttpMethod.Post, url)
                {
                    Content = content
                };
                request.Headers.Add("Authorization", $"Bearer {_config.ApiKey}");
                request.Headers.Add("User-Agent", "PacketPilot-Windows-Daemon/1.0");

                _logger.Debug("Registering apps", "url", url, "apps_count", apps.Count);

                using var response = await _httpClient.SendAsync(request, _cancellationTokenSource?.Token ?? CancellationToken.None);

                if (response.IsSuccessStatusCode)
                {
                    _logger.Debug("Apps registered successfully", "apps_count", apps.Count);
                }
                else
                {
                    var responseContent = await response.Content.ReadAsStringAsync();
                    _logger.Warn("Failed to register apps", "status", (int)response.StatusCode, "message", responseContent);
                }
            }
            catch (Exception ex)
            {
                _logger.Warn("Error registering apps", "error", ex.Message);
            }
        }

        private async Task SendPerProcessReportRequest(UsageReportRequest usageReport, string date, int appsCount, long totalRx, long totalTx)
        {
            var jsonData = JsonSerializer.Serialize(usageReport);

            var protocol = _config.UseTls ? "https" : "http";
            var url = $"{protocol}://{_config.ServerHost}:{_config.ServerPort}/api/v1/traffic/per-process";

            Exception? lastException = null;
            for (int attempt = 1; attempt <= _config.RetryAttempts; attempt++)
            {
                try
                {
                    using var content = new StringContent(jsonData, Encoding.UTF8, "application/json");
                    using var request = new HttpRequestMessage(HttpMethod.Post, url)
                    {
                        Content = content
                    };
                    request.Headers.Add("Authorization", $"Bearer {_config.ApiKey}");
                    request.Headers.Add("User-Agent", "PacketPilot-Windows-Daemon/1.0");

                    var totalRxMB = (double)totalRx / (1024 * 1024);
                    var totalTxMB = (double)totalTx / (1024 * 1024);

                    _logger.Debug("Sending daily usage report", "attempt", attempt, "url", url,
                        "date", date, "apps", appsCount, "total_rx_mb", totalRxMB, "total_tx_mb", totalTxMB);

                    using var response = await _httpClient.SendAsync(request, _cancellationTokenSource?.Token ?? CancellationToken.None);

                    if (response.IsSuccessStatusCode)
                    {
                        var responseContent = await response.Content.ReadAsStringAsync();
                        var serverResp = JsonSerializer.Deserialize<ServerResponse>(responseContent);

                        if (serverResp != null)
                        {
                            _logger.Info("Daily usage report sent successfully",
                                "message", serverResp.Message,
                                "date", date,
                                "apps", appsCount,
                                "total_rx_mb", totalRxMB,
                                "total_tx_mb", totalTxMB);
                        }

                        return;
                    }
                    else if (response.StatusCode == System.Net.HttpStatusCode.Forbidden)
                    {
                        // Device not activated - send health check to notify user
                        var responseContent = await response.Content.ReadAsStringAsync();
                        _logger.Warn("Device not activated - traffic report rejected",
                            "status", (int)response.StatusCode,
                            "message", responseContent);

                        await SendHealthCheckRequest();

                        lastException = new HttpRequestException($"Device not activated. Please wait for user approval.");
                        _logger.Info("Health check sent to notify user that device needs approval");
                    }
                    else
                    {
                        lastException = new HttpRequestException($"Server returned status {response.StatusCode}");
                        _logger.Warn("Server returned error status", "status", (int)response.StatusCode);
                    }
                }
                catch (Exception ex)
                {
                    lastException = ex;
                    _logger.Warn("Report attempt failed", "attempt", attempt, "error", ex.Message);
                }

                if (attempt < _config.RetryAttempts)
                {
                    var delay = UtilsHelper.ParseTimeSpan(_config.RetryDelay);
                    await Task.Delay(delay, _cancellationTokenSource?.Token ?? CancellationToken.None);
                }
            }

            if (lastException != null)
            {
                _logger.Error("All report attempts failed", "error", lastException.Message);
                throw lastException;
            }
        }

        private async Task SendTotalUsageReportRequest(TotalUsageReportRequest totalUsageReport, string date, long totalRx, long totalTx)
        {
            var jsonData = JsonSerializer.Serialize(totalUsageReport);

            var protocol = _config.UseTls ? "https" : "http";
            var url = $"{protocol}://{_config.ServerHost}:{_config.ServerPort}/api/v1/traffic/total-usage";

            Exception? lastException = null;
            for (int attempt = 1; attempt <= _config.RetryAttempts; attempt++)
            {
                try
                {
                    using var content = new StringContent(jsonData, Encoding.UTF8, "application/json");
                    using var request = new HttpRequestMessage(HttpMethod.Post, url)
                    {
                        Content = content
                    };
                    request.Headers.Add("Authorization", $"Bearer {_config.ApiKey}");
                    request.Headers.Add("User-Agent", "PacketPilot-Windows-Daemon/1.0");

                    var totalRxMB = (double)totalRx / (1024 * 1024);
                    var totalTxMB = (double)totalTx / (1024 * 1024);

                    _logger.Debug("Sending total usage report", "attempt", attempt, "url", url,
                        "date", date, "total_rx_mb", totalRxMB, "total_tx_mb", totalTxMB);

                    using var response = await _httpClient.SendAsync(request, _cancellationTokenSource?.Token ?? CancellationToken.None);

                    if (response.IsSuccessStatusCode)
                    {
                        var responseContent = await response.Content.ReadAsStringAsync();
                        var serverResp = JsonSerializer.Deserialize<ServerResponse>(responseContent);

                        if (serverResp != null)
                        {
                            _logger.Info("Total usage report sent successfully",
                                "message", serverResp.Message,
                                "date", date,
                                "total_rx_mb", totalRxMB,
                                "total_tx_mb", totalTxMB);
                        }

                        return;
                    }
                    else if (response.StatusCode == System.Net.HttpStatusCode.Forbidden)
                    {
                        var responseContent = await response.Content.ReadAsStringAsync();
                        _logger.Warn("Device not activated - total usage report rejected",
                            "status", (int)response.StatusCode,
                            "message", responseContent);

                        await SendHealthCheckRequest();

                        lastException = new HttpRequestException("Device not activated. Please wait for user approval.");
                        _logger.Info("Health check sent to notify user that device needs approval");
                    }
                    else
                    {
                        lastException = new HttpRequestException($"Server returned status {response.StatusCode}");
                        _logger.Warn("Server returned error status", "status", (int)response.StatusCode);
                    }
                }
                catch (Exception ex)
                {
                    lastException = ex;
                    _logger.Warn("Total usage report attempt failed", "attempt", attempt, "error", ex.Message);
                }

                if (attempt < _config.RetryAttempts)
                {
                    var delay = UtilsHelper.ParseTimeSpan(_config.RetryDelay);
                    await Task.Delay(delay, _cancellationTokenSource?.Token ?? CancellationToken.None);
                }
            }

            if (lastException != null)
            {
                _logger.Error("All total usage report attempts failed", "error", lastException.Message);
                throw lastException;
            }
        }

        public void Dispose()
        {
            _httpClient?.Dispose();
            _cancellationTokenSource?.Dispose();
        }
    }
}
